### PR TITLE
Cleanup

### PR DESCRIPTION
**Pull request by dorbarker**
The logic of the program was not altered, and so the output of the program should be identical. 
Feel free to ask if you have any questions or concerns with the changes.

Summary of changes:

- Simplify code in `run_INNUca()` and `main()`
    - Simplify adding results to `runs` dictionary
    - Move some step timing and progress printing ("RUNNING xxxx",  "END xxxx") to the `@timer` decorator in `utils.py`
    - The `run_report` values when a step was skipped are now taken from the variables `skipped` and `not_run`, rather than each being hardcoded individually
      
- Eliminate use of `nargs=1` in `utils.parseArguments()` which causes `ArgumentParser.add_argument()` to return a list of length 1 rather than the scalar that was intended

- Wrap report writing in functions and move them to `utils.py`

- Fixes bug with SPAdes kmer sizes that was causing crashes

- Fixes bug where an unsuccessful SPAdes run would cause the unbound variable `contigsFiltered` in `runSpades()` to be referenced, causing a crash

- Fixes bug where `--spadesKmers` stored user input as a comma-delimited `str` inside a `list` of length 1, but the default value was a list of `int`s
    - Both the user-provided value and the default value are now lists of `int`s

### CODE REVIEW

File: modules/utils.py [hunk 1]
```diff
@@ -282,7 +278,7 @@ def organizeSamplesFastq(directory, pairEnd_filesSeparation_list):
 		if len(samples[sample]) == 1:
 			print 'Only one fastq file was found: ' + str(samples[sample])
 			print 'Pair-End sequencing is required. This sample will be ignored'
-			samples.remove(sample)
+		        del samples[sample]
```
Comment (INNUENDOCON) [inline modules/utils.py]: @dorbarker: Since sample is a string, isn't it better to remove that element using .remove rather than del (which works with index)?

Comment (dorbarker) [inline modules/utils.py]: @INNUENDOCON  `samples` is a `dict`, which doesn't have the `.remove()` method. Removing elements from a `dict` is accomplished by `del`eting the key. This was causing a crash when innuca found an unpaired fastq. 

```python
>>> test_dict ={'a': 5, 'b': 6, 'c':42}
>>> test_dict.remove('a')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'dict' object has no attribute 'remove'

>>> del test_dict['a'] 
>>> print test_dict
{'c': 42, 'b': 6}
```
However, in writing this I just found a problem with my solution and I obviously forgot to test it properly - sorry! An iterable object like a `dict` cannot change size during iteration. 

We  would need something like this instead:
```python
samples_to_remove = []
for sample in samples:
    if len(samples[sample]) is 1:
        print 'Only one fastq file was found: ' + str(samples[sample])
        print 'Pair-End sequencing is required. This sample will be ignored'
        samples_to_remove.append(sample)

samples = {k: v for k, v in samples.items() if k not in samples_to_remove}
```

Comment (INNUENDOCON) [inline modules/utils.py]: Cool, thanks.
I forgot to mention that INNUENDOCON was me (@miguelpmachado).
BTW, I think I managed to understand the decorators and wraps. They are nice tools.
Thanks a lot.
Miguel

P.s.: I tried to make the code as clear as I could, but it's always hard to follow someone else code (at least for me), But I think you did a very nice job! It seems to me that you lost some parts of the code, but I'll confirm it, and I'll come to you with more news.

File: modules/utils.py [hunk 2]
```diff
@@ -373,3 +369,96 @@ def compressionType(file):
...
+			l = [run_report[step][0], run_report[step][2]]
+		else:
+			pass_qc = 'PASS' if run_report[step][1] else 'FAIL'
+			l = [run_report[step][0], pass_qc, run_report[step][2]]
+
+		line.extend(l)
+
+	return line
+
+def build_header():
```
Comment (INNUENDOCON) [inline modules/utils.py]: @dorbarker: When you did this, you missed the general sample report (same for write_sample_report function).
Another thing: you are only saving the samples_report file at the end of INNUca run.
So I changed it a little bit in order to keep the general sample report part and saving each time a sample ends so people can easily follow samples' run progression.
I hope you don't mind!
It will be something like this:
```
def start_sample_report_file(samples_report_path):
	header = ['#samples', 'samples_runSuccessfully', 'samples_passQC', 'samples_runningTime', 'samples_fileSize']
	for step in steps:
		if step == 'FastQ_Integrity':
			l = [step + '_filesOK', step + '_runningTime']
		elif step == 'Trimmomatic' or step == 'Pilon':
			l = [step + '_runSuccessfully', step + '_runningTime']
		else:
			l = [step + '_runSuccessfully', step + '_passQC', step + '_runningTime']
		header.extend(l)
	with open(samples_report_path, 'wt') as report:
		out = csv.writer(report, delimiter='\t')
		out.writerow(header)


def write_sample_report(samples_report_path, sample, run_successfully, pass_qc, runningTime, fileSize, run_report):
	line = [sample, run_successfully, 'PASS' if pass_qc else 'FAIL', runningTime, fileSize]
	line.extend(sampleReportLine(run_report))
	with open(samples_report_path, 'at') as report:
		out = csv.writer(report, delimiter='\t')
		out.writerow(line)
```
P.s.: I didn't checked it yet

File: INNUca.py [hunk 3]
```diff
@@ -218,221 +175,137 @@ def main():
...
+        # Remove Trimmomatic directory with cleaned reads
+	if not args.trimKeepFiles:
 
-	if runs['second_FastQC'][1] is False:
-		pass_qc = False
-	elif runs['second_FastQC'][1] is None and runs['first_FastQC'][1] is False:
-		pass_qc = False
+		# the except block formerly here never ran because error checking
+		# is performed internally by utils.removeDirectory()
+		utils.removeDirectory(trimmomatic_folder)
```
Comment (INNUENDOCON) [inline INNUca.py]: @dorbarker: the problem is with trimmomatic_folder variable that sometimes doesn't exist

File: INNUca.py [hunk 4]
```diff
@@ -218,221 +175,137 @@ def main():
...
 
-	if runs['SPAdes'][1] is False:
-		pass_qc = False
+	# Check run
+	run_successfully = all(runs[step][0] for step in runs if step != 'FastQ_Integrity')
 
-	if runs['MLST'][1] is False:
-		pass_qc = False
+	pass_qc = all([runs['FastQ_Integrity'][0], runs['first_Coverage'][1], runs['second_Coverage'][1],
```
Comment (INNUENDOCON) [inline INNUca.py]: @dorbarker: if I understand it right, in the cases of Coverage and FastQC checks, which have two runs, when the second time is None, pass_qc will be False. So, the original way should be more correct since it control for it. What do you think?

Comment (dorbarker) [inline INNUca.py]: @INNUENDOCON Yes, you are correct. I misunderstood, and thought both the first and second passes of FastQC and coverage needed to be true for `pass_qc` to be true.

Also, my original version introduced a bug where setting `--skipMLST` or `--skipSPAdes`, *i.e.* if `runs['MLST'][1]` or `runs['SPAdes'[1]` are equal to `None`, would cause the sample to fail QC. I didn't have enough coffee that day, apparently. 

Both bugs have been fixed in commit [9fdd5d8](https://github.com/INNUENDOCON/INNUca/pull/1/commits/9fdd5d85698d4850dc66703b16826edf925ef31a).

File: modules/utils.py (hunk 2 above)
Comment (dorbarker) [inline modules/utils.py]: I made writing all the sample reports at the end as a deliberate decision, but I was testing only ~10 genomes where seeing results in real time didn't matter. I agree that writing each sample as it is completed is the better approach, especially if you are running 100s or 1000s of genomes.
